### DISCUSSION TITLE
Different results with same model yoloV5s

### DISCUSSION THREAD

**Discussion by mgabell** [ultralytics/yolov5#10424] [General]
Hi,
I have used your repo for Yolov5 on Jetson Orin. I created a program around your repo and basically just added a consumer/producer pattern and XML output. 

Then I re-downloaded your repo and ran inference on two images using the test program.
**import torch
#Model
#model = torch.hub.load('Models/YoloV5', 'custom')  # or yolov5n - yolov5x6, custom
model = torch.hub.load('.', 'custom', 'best.pt', source='local')
#Images
#img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list
img = '/mnt/MyBook/Images/TestRun221129_Selection/ID0_AI_Image_104_1000508050.jpg'  # or file, Path, PIL, OpenCV, numpy, list
#Inference
results = model(img)
#Results
results.show() #print()  # or .show(), .save(), .crop(), .pandas(), etc.**

So I use same model as in "my" first program as in the latest downloaded repo. The result is different, in the "old" version I get a detection on both images, but on the latter I only get one hit.

Why is that? The same parameters is used, I did the same thing with the detect.py and 
all parameters should be the same, such as 

**conf_thres=0.25,  # confidence threshold
iou_thres=0.45,  # NMS IOU threshold
max_det=1000,  # maximum detections per image**

Any ideas? I can try to post the code here too. I need to clean it first :-|

I saw this, and will check it out but I use Pillow.Image to read the image..
https://github.com/ultralytics/yolov5/discussions/7627

Your code to read image:
**else:
Read image
self.count += 1
im0 = cv2.imread(path)  # BGR
assert im0 is not None, f'Image Not Found {path}'**

Whereas I 

`Image.open(fname)`

Any pointers to where to start looking would be appreciated. I am sorry for the formatting, its just not going my way..

**Comment by glenn-jocher** [comment]
👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Reply by mgabell** [reply]
Hi
I think, after some trial and error I found the issue.
One of the images is 1600x1300 px. The other is 1280x720.
When running the 1280x720 I add size to the inference:
results = model(img1, size=1280)
This yields the correct result where both images have the defect found.

I don't understand  why I need to specify the input height for the 1280x720 image.
Does it crop the image or rescale to 1280x1280?  I felt it should fix a re-scale to 720x720.

I dont know how to recreate the problem with your code without modification :-| I understand you will not look at my implementation of the full program, but I can rip out the part I think causing the issue.  I will create a small program doing what you do but with Image.open from PIL.

BTW
Also I think there is a problem with detect.py. If you are running that and you want to save the detected images, you get all images in the output folder. The save block is not indented to the right location? At-least for me I get all images to the output folder when choosing save-img=true.

**Reply by glenn-jocher** [reply]
@mgabell yes all images should be saved to the same directory, i.e. runs/detect/exp

**Reply by mgabell** [reply]
But that does not make sense. Why would I want a copy of all source files in the output folder? I want the files with detected objects?
I made changes myself to do that, but full copy was a bit strange in my opinion.

Just to take the opportunity to give some creds! Awsome job btw! thanks for this work. Really making AI obj. detection available to the masses..

**Reply by glenn-jocher** [reply]
@mgabell if you want to customize inference I'd suggest building your own workflows using python.

YOLOv5 🚀 [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36) models allow for simple model loading and inference in a pure python environment without using `detect.py`. 

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the YOLOv5 'small' model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints). Custom models can also be loaded, including custom trained PyTorch models and their [exported](https://github.com/ultralytics/yolov5/issues/251) variants, i.e. ONNX, TensorRT, TensorFlow, OpenVINO YOLOv5 models.

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # yolov5n - yolov5x6 official model
#                                            'custom', 'path/to/best.pt')  # custom model

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
results.xyxy[0]  # im predictions (tensor)

results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

results.pandas().xyxy[0].value_counts('name')  # class counts (pandas)
# person    2
# tie       1
```
<img width="550" src="https://user-images.githubusercontent.com/26833433/149679662-b2021c78-59b2-4f52-b768-1e5161537c85.png">

See [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36) for details.

Good luck 🍀 and let us know if you have any other questions!

**Reply by mgabell** [reply]
[](url)This is what I did, from a bit earlier version. My miss-understanding is the input size for the analysis in detect.py.
My image is 1280x720 and you specify the inference size [640x640]. As I read here:
https://github.com/ultralytics/yolov5/issues/2234
and here:
https://github.com/ultralytics/yolov3/issues/232
The image will be resized to meet minimum stride multiple of 32.

My apparent issue was that have 640 as image inference size for my image of 1280 seems to crop the image.
![image_issue](https://user-images.githubusercontent.com/78966452/206996834-f548b260-c6ba-4410-b4ea-0d928500b017.png)

As seen in the image, the black spot was detected when setting inference size to 640, but the clutter at the bottom was not.
Then I change the inference size to 1280 as the actual image size is then the clutter was found.

I will not claim scaling or cropping. Perhaps the re-scaling of the image made it impossible to detect. But if I have understood the inference size correctly I should get the full image analyzed with a 640 size but then its a matter of increasing the size incrementally until I get a suitable size for my particular use case

**Reply by glenn-jocher** [reply]
@mgabell we don't crop anything, images are resized and letterboxed (padded) to --imgsz.
